pkg/kubelet/kuberuntime/kuberuntime_sandbox.go

```diff
@@ -66,6 +66,12 @@ func (m *kubeGenericRuntimeManager) createPodSandbox(ctx context.Context, pod *v
 		}
 	}
 
+	// Set runtime handler annotation.
+	if podSandboxConfig.Annotations == nil {
+		podSandboxConfig.Annotations = make(map[string]string)
+	}
+	podSandboxConfig.Annotations["kubernetes.io/runtimehandler"] = runtimeHandler
```

Comment:
let's only add the annotations map and annotation if runtimeHandler != ""

Which can be done above after the log on line 65..

Reply:
I'm planning to use the presence of the annotation to determine how the cri server should behave.

For example, see this commit: https://github.com/microsoft/confidential-containers-containerd/commit/494207280f55c0ccd4c54ca7cfd24a8ae5d9fd09 (a variant of which I intend to include in a PR to containerd once we the kubelet starts adding the annotation).

The idea is that containerd would behave exactly as before without the annotation. If the annotation is present, even if runtimeHandler == "", it will return the image status based on the presence of a snapshot for the appropriate snapshotter.

This is something we discussed in that triage meeting some time back: we need the old behaviour when, for example, we want to do `crictl rmi` to remove an image from all snapshotters.

WDYT?

Reply:
https://github.com/microsoft/confidential-containers-containerd/commit/494207280f55c0ccd4c54ca7cfd24a8ae5d9fd09#r127070837

Reply:
if image.Annotations is == nil || runtimeHandler == "" then set runtimeHandler to c.config.ContainerdConfig.DefaultRuntimeName

Reply:
will this override the annotation if already exist?

Reply:
see link ^ ... here in kubelet you can leave annotations unset esp. when there are no annotations.. Over in your code just check for annotations being nil or if not nil then check for existence of the handler key/value pair

Reply:
In the beginning I had similar code to what you propose in the commit comment, however, that breaks `crictl rmi` and compatibility with kubelets that don't add runtime annotations.

The scenario is the following: you ran some container with a non-default runtime handler that has a different snapshotter, then you want to manually remove the image. If you run `crictl rmi X` or `crictl rmi -a`, it will fail because the snapshotter for the default runtime handler doesn't have a snapshot for the image (for it to work you'd have to add an annotation but the tool doesn't even support that yet for rmi like it does for pull).

Things also break in similar fashion if you pair this containerd with a kubelet that doesn't add runtime annotations: every time it tries to ask the status of an image for which the snapshotter of the default runtime doesn't have a snapshot, it will fail.

My solution to this issue, at a high level, is to preserve previous behaviour when the runtime handler annotation is absent, and only trigger new behaviour when it is present. This covers both scenarios above.

Reply:
agree with your logic.. LGTM

Reply:
@wedsonaf 
> will this override the annotation if already exist?

and should we avoid this?